sonar-text-plugin/src/main/java/org/sonar/plugins/secrets/api/task/ExecutorServiceManager.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * SonarQube Text Plugin
+ * Copyright (C) 2021-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.secrets.api.task;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ExecutorServiceManager {
+  private static final Logger LOG = LoggerFactory.getLogger(ExecutorServiceManager.class);
+  private ExecutorService lastExecutorService;
+
+  public ExecutorServiceManager() {
+    lastExecutorService = null;
+  }
+
+  private ExecutorService getLastExecutorService() {
+    // re-use last executor service if possible, otherwise allocate a new one
+    if (lastExecutorService == null || lastExecutorService.isShutdown()) {
+      lastExecutorService = Executors.newSingleThreadExecutor();
+    }
+    return lastExecutorService;
+  }
+
+  public boolean runWithTimeout(int timeoutMs, int ultimateTimeoutMs, Runnable run) {
+    ExecutorService executorService = getLastExecutorService();
+    Future<?> future = executorService.submit(run);
+
+    try {
+      future.get(timeoutMs, TimeUnit.MILLISECONDS);
+      return true;
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      executorService.shutdownNow();
```

Comment:
Why do we shutdown the whole service instead of only the thread?

Reply:
From my understanding, the `executorService` and the `Futur` object it return does not provide the capability to interact with the thread, so I cannot interrupt/shutdown it outside from the `executorService` (which is bad, because we have then to request a new thread when this happen instead of reusing the existing one).
Maybe I'm missing something?

Reply:
Both the `executorService` and the `Future` object are just asking for the thread to stop. Basically using `future.cancel(true);` will work exactly the same as `executorService.shutdownNow();`, but it is limited to the future itself.
Because we can't actively force the thread to stop, we are using the `InterruptibleCharSequence`. Here we add a way to interact with cancellation requests.

Reply:
I missed this method cancel(), let me change the logic and adapt to it, way better than shutting down the whole service!

Reply:
After some extensive search and testing on `future.cancel(true);`, this method requires the provided `Runnable` to support properly interruption mechanism.
If this is not the case, then it will "break" the `ExecutorService` object: it will still stop the current running thread, but any other provided task will neer be executed.
The worst is that we can't know when this happens: `future.cancel(true)` return true in both case, same for `future.isDone()` and `future.isCancelled()`.
No other method from ExecutorService provides any information about the status, calling `lastExecutorService.awaitTermination()` won't help because it only works as expected when the full executor service is being shut down.

As stated by the ticket:
> It should be noted, that the success of the interruption will be dependent on the underlying implementation of the regex matching algorithm (JDK dependent)

Because of this, there is no way to ensure that any `Runnable` provided will be interrupt-friendly, so unfortunately I don't think we can use this `future.cancel(true)` approach. With the ExecutorService, the only way is to keep the shutting down the whole ExecutorService way.

WDYT?